cmd/snap-update-ns/trespassing.go

```diff
@@ -0,0 +1,234 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017-2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"syscall"
+)
+
+// Assumptions track the assumptions about the state of the filesystem.
+//
+// Assumptions constitute the global part of the write restriction management.
+// Assumptions are global in the sense that they span multiple distinct write
+// operations. In contrast, Restrictions track per-operation state.
+type Assumptions struct {
+	unrestrictedPaths []string
+	pastChanges       []*Change
+}
+
+// AddUnrestrictedPaths adds a list of directories where writing is allowed
+// even if it would hit the real host filesystem (or transit through the host
+// filesystem). This is intended to be used with certain well-known locations
+// such as /tmp, $SNAP_DATA and $SNAP.
+func (as *Assumptions) AddUnrestrictedPaths(paths ...string) {
+	as.unrestrictedPaths = append(as.unrestrictedPaths, paths...)
+}
+
+// isRestricted returns true if a path follows restricted writing scheme.
+//
+// Writing to a restricted path results in step-by-step validation of each
+// directory, starting from the root of the file system. Unless writing is
+// allowed a mimic must be constructed to ensure that writes are not visible in
+// undesired locations of the host filesystem.
+//
+// Provided path is the full, absolute path of the entity that needs to be
+// created (directory, file or symbolic link).
+func (as *Assumptions) isRestricted(path string) bool {
+	// Anything rooted at one of the unrestricted paths is not restricted.
```

Comment:
Also, would it make sense to throw away paths whose prefix we already have right in the `AddUnrestrictedPaths`? e.g. AddUnrestrictedPaths("/foo", "/foo/bar") would only remember "/foo"? Or is it not worth optimizing?

Reply:
Right now it's not optimising. We are only adding three paths (hard-coded in main.go) that we allow.